### PR TITLE
Create kdump role

### PR DESCRIPTION
**Pull request by andrew103**
Wasn't sure whether or not to make a variable for the package version being used. I opted to not create a variable considering the state is set to `latest`.

### CODE REVIEW

File: AUTHORS [hunk 1]
```diff
@@ -0,0 +1 @@
+Author <authoremail@domain.net>
```
Comment (greg-hellings) [inline AUTHORS]: Please update this.

File: tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# role tasks
+- name: Upgrade kexec-tools to latest
+  yum:
```
Comment (greg-hellings) [inline tasks/main.yml]: Change this to "package" so it will work going forward into future systems that don't use `yum`.

File: tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    name: kexec-tools
+    state: latest
```
Comment (greg-hellings) [inline tasks/main.yml]: Change this to `state: present`. Setting this to `latest` is a hard no.

File: tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+  lineinfile: dest="{{ kdump_conf_path }}"
+              state=present
+              regexp='^(#?)dracut_args (.*)$'
+              line='dracut_args --mount "kdump.usersys.redhat.com:/var/www/html/vmcore /mnt/vmcore nfs defaults"'
+              backup=true
+  become: true
+  become_user: "{{ kdump_become_user }}"
+
+- name: Set relative path of vmcore folder
+  lineinfile: dest="{{ kdump_conf_path }}"
```
Comment (greg-hellings) [inline tasks/main.yml]: How large is the kdump.conf file? At a certain size it becomes preferable to run these lineinfile commands as a single `template` update. Alternatively, can we look at combining all three of these into a single call using a loop?

File: tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+              state=present
+              regexp='^path (.*)$'
+              line='path /{{inventory_hostname}}'
+              backup=true
+  become: true
+  become_user: "{{ kdump_become_user }}"
+  register: kdump_conf
+
+- name: Enable kdump service
+  systemd:
```
Comment (greg-hellings) [inline tasks/main.yml]: You don't have a `state: started` here. Is that intentional?

Additionally, let's switch to using the `service` module where possible. I understand it doesn't have the `masked` option, but I'm not sure that's necessary. If it is, then let's push the `enabled: true` down to a following task and use `service` there.

File: tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    name: kdump
+    enabled: true
+    masked: false
+  become: true
+  become_user: "{{ kdump_become_user }}"
+
+- name: Restart kdump service
+  systemd:
+    state: restarted
```
Comment (greg-hellings) [inline tasks/main.yml]: Why is `restarted` in the main playbook and not in the `handlers.yml` file?

**Review by greg-hellings — CHANGES_REQUESTED** [review]

File: tasks/main.yml (hunk 4 above)
Comment (andrew103) [reply tasks/main.yml]: The file is a little over 160 lines long

File: tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,52 @@
+# role tasks
+- name: Upgrade kexec-tools to latest
+  yum:
+    name: kexec-tools
+    state: latest
+  become: true
+  become_user: "{{ kdump_become_user }}"
+
+- name: Waive previous nfs setup if enabled
+  lineinfile: dest="{{ kdump_conf_path }}"
```
Comment (robled) [inline tasks/main.yml]: When writing Ansible YAML we should always use colons instead of equals signs, with two-space indentation.  Example:
```
lineinfile: 
  dest: "{{ kdump_conf_path }}"
  state: present
```

File: tasks/main.yml [hunk 8]
```diff
@@ -1 +1,23 @@
 # role tasks
+- name: Upgrade kexec-tools to latest
```
Comment (greg-hellings) [inline tasks/main.yml]: This string should be updated to not mention "latest"

File: tasks/main.yml [hunk 9]
```diff
@@ -1 +1,23 @@
...
+  package:
+    name: "{{ kdump_version }}"
+    state: present
+  become: true
+  become_user: "{{ kdump_become_user }}"
+
+- name: Edit kdump config file
+  template:
+    src: ../templates/kdump_conf
```
Comment (greg-hellings) [inline tasks/main.yml]: You can remove the `../templates/` from this as, by default, the `template` module will look in the `templates` folder of the current role.

File: defaults/main.yml [hunk 10]
```diff
@@ -1 +1,4 @@
 # defaults file for setup_kdump
+kdump_version: kexec-tools
```
Comment (greg-hellings) [inline defaults/main.yml]: Is this really a "version" or is it just the RPM package name?

File: .travis.yml [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+  - pip --version
+  - pip install ansible ansible-lint yamllint flake8 molecule docker
+  - ansible --version
+  - ansible-lint --version
+  - yamllint --version
+  - flake8 --version
+  - molecule --version
+
+script:
+  - molecule lint -s openstack
```
Comment (greg-hellings) [inline .travis.yml:18]: Why did you remove the `docker` test? I'm guessing there are errors with the service startup in that environment? In that case, let's remove the `molecule/docker` directory as well.

File: templates/kdump_conf [hunk 12]
```diff
@@ -0,0 +1,3 @@
+path /{{inventory_hostname}}
```
Comment (greg-hellings) [inline templates/kdump_conf]: If the target name for this file is `kdump.conf` then it should be named `kdump.conf` here, as well instead of `kdump_conf`. Otherwise, you lose any potential for an editor to properly grok the syntax highlighting.

**Review by greg-hellings — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 11 above)
Comment (andrew103) [reply .travis.yml:18]: Docker just wasn't able to find the package that needed to be installed. I went ahead and removed the `molecule/docker` directory

File: templates/kdump.conf [hunk 13]
```diff
@@ -0,0 +1,3 @@
+path /{{inventory_hostname}}
+core_collector makedumpfile -l --message-level 1 -d 31
+dracut_args --mount "kdump.usersys.redhat.com:/var/www/html/vmcore /mnt/vmcore nfs defaults"
```
Comment (robled) [inline templates/kdump.conf]: `core_collector` and `dracut_args` look like site-specific configuration that should be parameterized with sensible defaults.

**Review by robled — CHANGES_REQUESTED** [review]

File: templates/kdump.conf (hunk 13 above)
Comment (greg-hellings) [reply templates/kdump.conf]: Also, line 1 probably should be configurable in case the user has an `inventory_hostname` that isn't as fun for pathnames as well.

File: meta/main.yml [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+  company: Red Hat, Inc.
+  license: GPLv3
+  min_ansible_version: 2.4
+  platforms:
+    - name: EL
+      versions:
+        - 7
+
+  galaxy_tags:
+    - oasis
```
Comment (greg-hellings) [inline meta/main.yml:13]: I'm not sure why you pulled out
```yaml
dependencies: []
```
from here, but it's considered best practices to leave that in and have it always be blank.

File: defaults/main.yml [hunk 15]
```diff
@@ -0,0 +1,8 @@
+# defaults file for setup_kdump
+kdump_package: kexec-tools
+kdump_conf_path: /etc/kdump.conf
+kdump_become_user: root
+
+kdump_hostname: kdump-setup-service
+kdump_core_collector_args: makedumpfile -l --message-level 1 -d 31
+kdump_dracut_mnt_pth: kdump.usersys.redhat.com:/var/www/html/vmcore /mnt/vmcore
```
Comment (robled) [inline defaults/main.yml]: This would depend on having an NFS server running somewhere on the network.  I suspect that many use cases would not involve an NFS server.  I think we may need to point to a local path for this, or even omit it, depending on how this option behaves by default.  We'll have to learn more about how kdump works to determine the most sensible default here.

File: templates/kdump.conf [hunk 16]
```diff
@@ -0,0 +1,3 @@
+path /{{ kdump_hostname }}
+core_collector {{kdump_core_collector_args}}
+dracut_args --mount "{{kdump_dracut_mnt_pth}} nfs defaults"
```
Comment (robled) [inline templates/kdump.conf]: Related to my previous comment - NFS should probably be entirely optional here.

**Review by robled — CHANGES_REQUESTED** [review]

File: meta/main.yml (hunk 14 above)
Comment (andrew103) [reply meta/main.yml:13]: Added it back

File: tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,23 @@
...
+    dest: "{{ kdump_conf_path }}"
+  become: true
+  become_user: "{{ kdump_become_user }}"
+  notify:
+    - Restart kdump service
+
+- name: Enable kdump service
+  service:
+    name: kdump
+    state: started
```
Comment (robled) [inline tasks/main.yml:21]: I think we need an `enabled: true` in here as well to ensure that the service will get started after a reboot.

File: defaults/main.yml [hunk 18]
```diff
@@ -0,0 +1,16 @@
+# defaults file for setup_kdump
+kdump_package: kexec-tools
+kdump_conf_path: /etc/kdump.conf
+kdump_become_user: root
+
+kdump_set_crash_path: true
+kdump_crash_path: var/crash
```
Comment (robled) [inline defaults/main.yml]: This path isn't valid.  It should be `/var/crash`.

File: defaults/main.yml [hunk 19]
```diff
@@ -0,0 +1,16 @@
...
+
+kdump_set_core_collector: false
```
Comment (robled) [inline defaults/main.yml]: The user interface would be simpler if we could just check if the var is defined to populate the template with the arguments for all settings in this file.  @greg-hellings - Is our best practice to do this the way we did it in cinch? [1][2].  

In that case we left the vars commented out in _defaults/main.yml_ (but documented in the README.md) and then used an `omit` filter in the tasks.

- [1] https://github.com/RedHatQE/cinch/blob/master/cinch/roles/jenkins_common/defaults/main.yml#L8
- [2] https://github.com/RedHatQE/cinch/blob/master/cinch/roles/jenkins_common/tasks/main.yml#L35

**Review by robled — CHANGES_REQUESTED** [review]

File: defaults/main.yml (hunk 19 above)
Comment (greg-hellings) [reply defaults/main.yml]: Indeed it is, although you'd have to do it a little differently since this value is consumed in a template instead of in an args. So `| default(omit)` isn't viable. But you can still check for the value being defined at all in the template.

File: templates/kdump.conf [hunk 20]
```diff
@@ -0,0 +1,13 @@
+# kdump configuration file
+
+{% if kdump_set_crash_path -%}
+path /{{ kdump_crash_path }}
+{%- endif %}
+
+{% if kdump_set_core_collector -%}
```
Comment (greg-hellings) [inline templates/kdump.conf]: Here you should be able to do something like `{% if kdump_core_collector_args is defined -%}`. At least I think that's the appropriate syntax.

File: README.md [hunk 21]
```diff
@@ -1 +1,56 @@
...
+
+None
+
+Example Playbook
+----------------
+
+```
+- hosts: setup_kdump-servers
+  roles:
+    - kdump
```
Comment (robled) [inline README.md]: In here it may be helpful to have some examples of the variables being used ([example](https://github.com/oasis-roles/jbosseap)) - with actual working values.  Rather than saying "commented out by default" we could clarify by saying "not configured by default".

File: defaults/main.yml [hunk 22]
```diff
@@ -0,0 +1,13 @@
+# defaults file for setup_kdump
+kdump_package: kexec-tools
+kdump_conf_path: /etc/kdump.conf
+kdump_become_user: root
+
+kdump_crash_path: /var/crash
+
+# kdump_core_collector_args: "{{ omit }}"
```
Comment (robled) [inline defaults/main.yml]: Since these are commented out, we should change the values from `{{ omit }}` to something helpful to look at documentation-wise.  In the end, the example values we provide should be a full working setup with all the optional vars being used.  Example:

https://github.com/RedHatQE/cinch/blob/master/cinch/roles/jenkins_common/defaults/main.yml#L8

File: tasks/main.yml (hunk 17 above)
Comment (andrew103) [reply tasks/main.yml:21]: Added

**Review by robled — CHANGES_REQUESTED** [review]

**Review by robled — APPROVED** [review]

File: vars/main.yml [hunk 23]
```diff
@@ -0,0 +1 @@
+# Vars for setup_kdump
```
Comment (greg-hellings) [inline vars/main.yml]: Since this isn't in use, go ahead and delete it.

File: templates/kdump.conf [hunk 24]
```diff
@@ -0,0 +1,13 @@
+# kdump configuration file
+
+{% if kdump_crash_path is defined -%}
+path {{ kdump_crash_path }}
+{%- endif %}
+
+{% if kdump_core_collector_args is defined -%}
+core_collector {{kdump_core_collector_args}}
```
Comment (greg-hellings) [inline templates/kdump.conf]: Your templating syntax is inconsistent. In other places in the role you have a space between `{{` and the variable name. Here and on line 12 you don't. Our preference is to have the spaces for the ease of reading, but at the least keep it consistent within a role.

**Review by greg-hellings — CHANGES_REQUESTED** [review]

**Review by greg-hellings — APPROVED** [review]
